### PR TITLE
Support convert LoRA safetensors into diffusers format

### PR DESCRIPTION
**Pull request by haofanwang**
This PR hanldes https://github.com/huggingface/diffusers/issues/2363 and part of https://github.com/huggingface/diffusers/issues/2326.

In summary, given a LoRA weight in safetensors format, we can merge it into a stable diffusion model in diffusers. It is quite easy to use as below. I have tested it with civitai models.

`python convert_lora_safetensor_to_diffusers.py --base_model_path runwayml/stable-diffusion-v1-5 --checkpoint_path  xxx.safetensors --dump_path ./outputs`

### CODE REVIEW

**Comment by haofanwang** [comment]
I have run quality checker on locally, it should be fine to merge.

**Comment by haofanwang** [comment]
@garyhxfang Yes, I did try this using two different LoRA (as style insertion), it would work.

**Comment by jndietz** [comment]
@haofanwang This is awesome.

This is kind of a loaded question, and not really related to the PR, but how did you "know" what to code for this PR?  How can I learn more about the structure of pytorch, safetensors, and diffusers binaries and use them effectively?

**Comment by garyhxfang** [comment]
@haofanwang  Hi, I try the script, and it work well currently. Thanks a lot for your great work!
Only a small mistake
`parser.add_argument("--alpha", default=0.75, type=int, help="The merging ratio in W = W0 + alpha * deltaW") `
Think the type of alpha should be float? I can run after I change it to float.

**Comment by haofanwang** [comment]
I'm glad to know it is helpful to you! I have fixed this typo, thanks.

File: scripts/convert_lora_safetensor_to_diffusers.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            pair_keys.append(key)
+        else:
+            pair_keys.append(key)
+            pair_keys.append(key.replace("lora_up", "lora_down"))
+
+        # update weight
+        if len(state_dict[pair_keys[0]].shape) == 4:
+            weight_up = state_dict[pair_keys[0]].squeeze(3).squeeze(2).to(torch.float32)
+            weight_down = state_dict[pair_keys[1]].squeeze(3).squeeze(2).to(torch.float32)
+            curr_layer.weight.data += alpha * torch.mm(weight_up, weight_down).unsqueeze(2).unsqueeze(3)
```
Comment (patrickvonplaten) [inline scripts/convert_lora_safetensor_to_diffusers.py:80]: I don't fully understand this - why do are we updating the layer weight here?

Comment (patrickvonplaten) [reply scripts/convert_lora_safetensor_to_diffusers.py:80]: Ah I see! Smart! Ok cool, let's go for it

**Comment by pythonlearner1025** [comment]
> 

https://haofanwang.github.io/

**Comment by AlexxxStarkkk** [comment]
I used the “convert_lora_safetensor_to_diffusers.py” to convert the [one-piece-wano-saga-style-lora]--“wanostyle_2_offset.safesensor” to a folder(size about 5G)。
then, i used it:
`import torch, os
from diffusers import StableDiffusionPipeline, DPMSolverMultistepScheduler
from diffusers.models import AutoencoderKL

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

model_id = r".\sd_model\stable-diffusion-v1-5"  # local sd model
vae = AutoencoderKL.from_pretrained("stabilityai/sd-vae-ft-mse", cache_dir=r"D:\sd310\vae_model")

pipe = StableDiffusionPipeline.from_pretrained(model_id)

lora_model_path = r'D:\sd310\lora_model\diffusers_format\one_piece_diffusers_model'
pipe.unet.load_attn_procs(lora_model_path)

pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config)

pipe = pipe.to("cuda")

prompt = "1boy, wanostyle, monkey d luffy, smiling"

image = pipe(prompt, height=504, width=680).images[0]  # PIL格式 (https://pillow.readthedocs.io/en/stable/)
image.save("./001_images_test/one_piece1.png".format(prompt[0:10]))`

but there is an error:
`Traceback (most recent call last):
  File "D:\sd310\main_test.py", line 69, in <module>
    pipe.unet.load_attn_procs(lora_model_path)
  File "E:\programs\Anoconda3\envs\sd310\lib\site-packages\diffusers\loaders.py", line 204, in load_attn_procs
    model_file = _get_model_file(
  File "E:\programs\Anoconda3\envs\sd310\lib\site-packages\diffusers\utils\hub_utils.py", line 275, in _get_model_file
    raise EnvironmentError(
OSError: Error no file named pytorch_lora_weights.bin found in directory D:\sd310\lora_model\diffusers_format\one_piece_diffusers_model.`

how can i fix it?

**Pull request merged**